src/lua/serializer.c

```diff
@@ -88,16 +89,139 @@ luaL_serializer_create(struct luaL_serializer *cfg)
 {
 	rlist_create(&cfg->on_update);
 	for (int i = 0; OPTIONS[i].name != NULL; i++) {
-		int *pval = (int *) ((char *) cfg + OPTIONS[i].offset);
-		*pval = OPTIONS[i].defvalue;
+		switch (OPTIONS[i].type) {
+		case LUA_TBOOLEAN:
+		case LUA_TNUMBER: {
+			int *pval = (int *)((char *)cfg + OPTIONS[i].offset);
+			*pval = OPTIONS[i].defvalue;
+			break;
+		}
+		case LUA_TTABLE: {
+			if (strcmp(OPTIONS[i].name, "encode_key_order") != 0) {
+				unreachable();
+			}
+
+			char **p = (char **)((char *)cfg + OPTIONS[i].offset);
+			*p = NULL;
+			break;
+		}
+		default:
+			unreachable();
+		}
 	}
 }
 
 void
 luaL_serializer_copy_options(struct luaL_serializer *dst,
 			     const struct luaL_serializer *src)
 {
+	luaL_serializer_options_delete(dst);
 	memcpy(dst, src, offsetof(struct luaL_serializer, end_of_options));
+
+	/*
+	 * Do a deep copy of the encoder key order. Not doing so would lead to
+	 * double free.
+	 */
+	size_t encode_key_order_len = 0;
+	while (src->encode_key_order != NULL &&
+	       src->encode_key_order[encode_key_order_len] != NULL) {
+		encode_key_order_len++;
+	}
+
+	if (encode_key_order_len > 0) {
+		dst->encode_key_order = xcalloc(
+			encode_key_order_len + 1, sizeof(char *));
+
+		for (size_t i = 0; i < encode_key_order_len; i++) {
+			size_t len = strlen(src->encode_key_order[i]);
+			dst->encode_key_order[i] = xcalloc(len + 1,
+							   sizeof(char));
+			memcpy(dst->encode_key_order[i],
+			       src->encode_key_order[i],
+			       len);
+		}
+	} else {
+		dst->encode_key_order = NULL;
+	}
+}
+
+/**
+ * Free memory allocated for encode_key_order.
+ * @param cfg Serializer configuration.
+ */
+static void
+luaL_serializer_free_encode_key_order(struct luaL_serializer *cfg)
+{
+	if (cfg->encode_key_order == NULL) {
+		return;
+	}
+
+	for (size_t i = 0; cfg->encode_key_order[i] != NULL; i++) {
+		free(cfg->encode_key_order[i]);
+	}
+
+	free(cfg->encode_key_order);
+	cfg->encode_key_order = NULL;
+}
+
+void
+luaL_serializer_options_delete(struct luaL_serializer *cfg)
+{
+	assert(cfg);
+	luaL_serializer_free_encode_key_order(cfg);
+	/* This is a partial TRASH(cfg), only for options. */
+	memset(cfg, '#', offsetof(struct luaL_serializer, end_of_options));
+}
+
+/**
+ * Destroy the serializer object, freeing all allocated memory.
+ * @param L Lua stack.
+ * @retval 0.
+ */
+static int
+luaL_destroy_serializer(struct lua_State *L)
```

Comment:
Nit: All the other functions start from `luaL_serializer_`.

Reply:
Fixed.